src/Simplex/Chat.hs

```diff
@@ -219,11 +223,52 @@ processAgentMessage User {userId, profile} agentConnId agentMessage = do
         CONF confId connInfo -> do
           -- TODO update connection status
           saveConnInfo conn connInfo
-          withAgent $ \a -> allowConnection a agentConnId confId $ encodeProfile profile
+          withAgent $ \a -> allowConnection a agentConnId confId . directMessage $ XInfo profile
         INFO connInfo ->
           saveConnInfo conn connInfo
         _ -> pure ()
-    _ -> pure ()
+    ReceivedGroupMessage gName m ->
+      case agentMessage of
+        CONF confId connInfo -> do
+          ChatMessage {chatMsgEvent} <- liftEither $ parseChatMessage connInfo
+          case chatMsgEvent of
+            XGrpAcpt memId
+              | memId == memberId m -> do
+                withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemAccepted
+                withAgent $ \a -> allowConnection a agentConnId confId ""
+              | otherwise -> pure () -- TODO error not matching member ID
+            _ -> pure () -- TODO show/log error, other events in SMP confirmation
+        CON -> do
+          Group {membership, members} <- withStore $ \st -> getGroup st user gName
+          -- TODO because the contact is not created instantly, if the member is not created from contact,
+          -- it should still have a unique local display name.
+          -- If it is created from contact it can still be duplicated on the member (and match the contact)
+          case invitedBy m of
+            IBUser -> do
+              -- sender was invited by the current user
+              withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemConnected
+              sendGroupMessage members $ XGrpMemNew (memberId m) (memberRole m) (memberProfile m)
+              showConnectedGroupMember gName $ displayName (memberProfile m :: Profile)
+              forM_ (filter (\m' -> memberStatus m' >= GSMemConnected) . map fst $ members) $ \m' ->
+                sendDirectMessage agentConnId $ XGrpMemIntro (memberId m') (memberRole m') (memberProfile m')
+            _ -> do
+              if Just (invitedBy membership) == (IBContact <$> memberContactId m)
+                then do
+                  -- sender invited the current user
+                  withStore $ \st -> updateGroupMemberStatus st userId (groupMemberId m) GSMemConnected
+                  showUserConnectedToGroup gName
+                  pure ()
+                else do
+                  showConnectedGroupMember gName $ displayName (memberProfile m :: Profile)
+        MSG meta msgBody -> do
+          ChatMessage {chatMsgEvent} <- liftEither $ parseChatMessage msgBody
+          case chatMsgEvent of
+            XGrpMemNew memId memRole memProfile -> do
+              Group {membership, members} <- withStore $ \st -> getGroup st user gName
+              when (memberId membership /= memId && isNothing (find ((== memId) . memberId . fst) members)) $
+                withStore $ \st -> pure () -- add new member as GSMemAccepted
```

Comment:
you're probably doing it in next pr, but I'll comment here just in case - we should check for status of the inviter here (whether he's at least admin), because check on the invitee part is not enough since both he and the inviter can be compromised / faulty

Reply:
yes - this is not complete